src/transformers/modeling_utils.py

```diff
@@ -849,21 +846,40 @@ def generate(
 
         if self.config.is_encoder_decoder:
             assert bos_token_id is not None, "Encoder Decoder Models need to have a bos_token_id"
-            # encoder decoder need to start with empty input_ids and copy the input_ids to encoder_inputs
-            encoder_inputs = input_ids
+
+            # only need to generate encoder_outputs once for encoder-decoder
+            if hasattr(self, "encoder"):
+                encoder = self.encoder
+            elif hasattr(self, "model") and hasattr(self.model, "encoder"):
+                encoder = self.model.encoder
+            else:
+                raise NotImplementedError("{} does not seem to have an encoder".format(self))
+
+            # T5 needs to convert to embeddings first
+            if hasattr(self, "shared"):
```

Comment:
`BartModel` (not `BartForConditionalGeneration`) has `self.shared` and unit tests never generate with BartModel, so this will break that I think.
Is there a way to keep t5 specific logic in modeling_t5.py?

Reply:
Yeah, I agree this is not ideal here. I mean BartModel would never `generate()` because it does not have an `lm_head`, but I think you're right this logic should be in each of the different model files. So we could introduce a standard for encoder-decoder with `lm_head` models to have a `encode_forward` function. I would be fine with that @thomwolf @craffel.